controller/work_item_events.go

```diff
@@ -41,165 +44,181 @@ func (c *EventsController) List(ctx *app.ListWorkItemEventsContext) error {
+			WorkItemType: &app.RelationGeneric{
+				Links: &app.GenericLinks{
+					Self: ptr.String(rest.AbsoluteURL(req, app.WorkitemtypeHref(wit.ID))),
 				},
-				NewValue: &app.RelationGenericList{
-					Data: []*app.GenericData{new},
+				Data: &app.GenericData{
+					ID:   ptr.String(wit.ID.String()),
+					Type: ptr.String(APIStringTypeWorkItemType),
 				},
 			},
+		},
+	}
+
+	convertVal := func(kind workitem.Kind, val interface{}) (interface{}, bool) {
+		switch kind {
+		case workitem.KindString,
+			workitem.KindInteger,
+			workitem.KindFloat,
+			workitem.KindBoolean,
+			workitem.KindURL,
+			workitem.KindMarkup,
+			workitem.KindDuration, // TODO(kwk): get rid of duration
+			workitem.KindInstant:
+			return val, false
+		case workitem.KindIteration:
+			data, _ := ConvertIterationSimple(req, val)
+			return data, true
+		case workitem.KindUser:
+			data, _ := ConvertUserSimple(req, val)
+			return data, true
+		case workitem.KindLabel:
+			data := ConvertLabelSimple(req, val)
+			return data, true
+		case workitem.KindBoardColumn:
+			data := ConvertBoardColumnSimple(req, val)
+			return data, true
+		case workitem.KindArea:
+			data, _ := ConvertAreaSimple(req, val)
+			return data, true
+		case workitem.KindCodebase:
+			data, _ := ConvertCodebaseSimple(req, val)
+			return data, true
 		}
-	case workitem.SystemIteration:
-		old, _ := ConvertIterationSimple(request, wiEvent.Old)
-		new, _ := ConvertIterationSimple(request, wiEvent.New)
-		e = &app.Event{
-			Type: event.APIStringTypeEvents,
-			ID:   &wiEvent.ID,
-			Attributes: map[string]interface{}{
-				"name":      wiEvent.Name,
-				"newValue":  nil,
-				"oldValue":  nil,
-				"timestamp": wiEvent.Timestamp,
-			},
+		return nil, false
+	}
 
-			Relationships: &app.EventRelations{
-				Modifier: modifier,
-				OldValue: &app.RelationGenericList{
-					Data: []*app.GenericData{old},
-				},
-				NewValue: &app.RelationGenericList{
-					Data: []*app.GenericData{new},
-				},
-			},
+	kind := fieldDef.Type.GetKind()
+	if kind == workitem.KindEnum {
+		enumType, ok := fieldDef.Type.(workitem.EnumType)
+		if !ok {
+			return nil, errs.Errorf("failed to convert field \"%s\" to enum type: %+v", fieldName, fieldDef)
 		}
-	case workitem.SystemAssignees:
-		e = &app.Event{
-			Type: event.APIStringTypeEvents,
-			ID:   &wiEvent.ID,
-			Attributes: map[string]interface{}{
-				"name":      wiEvent.Name,
-				"newValue":  nil,
-				"oldValue":  nil,
-				"timestamp": wiEvent.Timestamp,
-			},
-			Relationships: &app.EventRelations{
-				Modifier: modifier,
-				OldValue: &app.RelationGenericList{
-					Data: ConvertUsersSimple(request, wiEvent.Old.([]interface{})),
-				},
-				NewValue: &app.RelationGenericList{
-					Data: ConvertUsersSimple(request, wiEvent.New.([]interface{})),
-				},
-			},
+		kind = enumType.BaseType.GetKind()
+	}
+
+	// handle all single value fields (including enums)
+	if kind != workitem.KindList {
+		oldVal, useRel := convertVal(kind, wiEvent.Old)
```

Comment:
Oh, if it is `true` it means that a relationship has to be used to store the value; otherwise we use the `oldValue` and `newValue` members from the `attributes`.